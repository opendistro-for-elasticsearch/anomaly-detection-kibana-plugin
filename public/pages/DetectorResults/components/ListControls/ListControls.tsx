/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { EuiFlexGroup, EuiFlexItem, EuiPagination } from '@elastic/eui';

interface ListControlsProps {
  activePage: number;
  pageCount: number;
  onPageClick: (pageNumber: number) => void;
}
export const ListControls = (props: ListControlsProps) => (
  <EuiFlexGroup
    style={{ padding: '0px 5px' }}
    alignItems="center"
    justifyContent="flexEnd"
  >
    {props.pageCount > 1 ? (
      <EuiFlexItem grow={false} style={{ justifyContent: 'center' }}>
        <EuiPagination
          pageCount={props.pageCount}
          activePage={props.activePage}
          onPageClick={props.onPageClick}
          data-test-subj="anomaliesPageControls"
        />
      </EuiFlexItem>
    ) : null}
  </EuiFlexGroup>
);
