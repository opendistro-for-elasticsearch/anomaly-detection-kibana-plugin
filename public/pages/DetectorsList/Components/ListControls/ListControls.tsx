/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiComboBox,
  EuiComboBoxOptionProps,
  EuiFieldSearch,
  EuiFlexGroup,
  EuiFlexItem,
  EuiPagination,
} from '@elastic/eui';
import React from 'react';
import { getDetectorStateOptions } from '../../utils/helpers';
import { DETECTOR_STATE } from '../../../../utils/constants';

interface ListControlsProps {
  activePage: number;
  pageCount: number;
  search: string;
  selectedDetectorStates: DETECTOR_STATE[];
  selectedIndices: string[];
  indexOptions: EuiComboBoxOptionProps[];
  onDetectorStateChange: (options: EuiComboBoxOptionProps[]) => void;
  onIndexChange: (options: EuiComboBoxOptionProps[]) => void;
  onSearchDetectorChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onSearchIndexChange: (searchValue: string) => void;
  onPageClick: (pageNumber: number) => void;
}
export const ListControls = (props: ListControlsProps) => (
  <EuiFlexGroup gutterSize="s">
    <EuiFlexItem grow={false} style={{ width: '40%' }}>
      <EuiFieldSearch
        fullWidth={true}
        value={props.search}
        placeholder="Search"
        onChange={props.onSearchDetectorChange}
        data-test-subj="detectorListSearch"
      />
    </EuiFlexItem>
    <EuiFlexItem>
      <EuiComboBox
        id="selectedDetectorStates"
        placeholder="All detector states"
        isClearable={true}
        singleSelection={false}
        options={getDetectorStateOptions()}
        onChange={props.onDetectorStateChange}
        selectedOptions={
          props.selectedDetectorStates.length > 0
            ? props.selectedDetectorStates.map(index => ({ label: index }))
            : []
        }
      />
    </EuiFlexItem>
    <EuiFlexItem>
      <EuiComboBox
        id="selectedIndices"
        placeholder="All indices"
        isClearable={true}
        singleSelection={false}
        options={props.indexOptions}
        onChange={props.onIndexChange}
        onSearchChange={props.onSearchIndexChange}
        selectedOptions={
          props.selectedIndices.length > 0
            ? props.selectedIndices.map(index => ({ label: index }))
            : []
        }
      />
    </EuiFlexItem>
    {props.pageCount > 1 ? (
      <EuiFlexItem grow={false} style={{ justifyContent: 'center' }}>
        <EuiPagination
          pageCount={props.pageCount}
          activePage={props.activePage}
          onPageClick={props.onPageClick}
          data-test-subj="detectorPageControls"
        />
      </EuiFlexItem>
    ) : null}
  </EuiFlexGroup>
);
