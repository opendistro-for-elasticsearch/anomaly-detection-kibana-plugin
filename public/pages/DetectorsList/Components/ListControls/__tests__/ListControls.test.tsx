/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import {
  ALL_DETECTOR_STATES,
  ALL_INDICES,
  DETECTOR_STATE,
} from '../../../../utils/constants';
import { ListControls } from '../ListControls';

describe('<ListControls /> spec', () => {
  const defaultProps = {
    activePage: 1,
    pageCount: 10,
    search: '',
    selectedDetectorStates: ALL_DETECTOR_STATES,
    selectedIndices: ALL_INDICES,
    detectorStateOptions: [],
    indexOptions: [],
    onDetectorStateChange: jest.fn(),
    onIndexChange: jest.fn(),
    onSearchDetectorChange: jest.fn(),
    onSearchIndexChange: jest.fn(),
    onPageClick: jest.fn(),
  };
  beforeEach(() => {
    jest.clearAllMocks();
  });
  describe('Empty results', () => {
    test('renders component with empty message', async () => {
      const { container } = render(<ListControls {...defaultProps} />);
      expect(container.firstChild).toMatchSnapshot();
    });
    test('should call onSearchDetectorChange callback when user inputs text', () => {
      const { getByPlaceholderText } = render(
        <ListControls {...defaultProps} />
      );
      userEvent.type(getByPlaceholderText('Search'), 'Testing');
      expect(defaultProps.onSearchDetectorChange).toHaveBeenCalledTimes(7);
    });
    test('pagination should be hidden if pages count is 1', async () => {
      const { queryByTestId } = render(
        <ListControls {...defaultProps} pageCount={1} />
      );
      expect(queryByTestId('detectorPageControls')).toBeNull();
    });
    test('pagination should be visible if pages count more than 1', async () => {
      const { queryByTestId } = render(
        <ListControls {...defaultProps} pageCount={2} />
      );
      expect(queryByTestId('detectorPageControls')).not.toBeNull();
    });
    test('should call onPageClick when click on the page', async () => {
      const { getByTestId } = render(<ListControls {...defaultProps} />);
      fireEvent.click(getByTestId('pagination-button-3'));
      expect(defaultProps.onPageClick).toHaveBeenCalledTimes(1);
    });
    test('should display default detector state and index options', () => {
      const { getByText } = render(<ListControls {...defaultProps} />);
      expect(getByText('All detector states')).toBeInTheDocument();
      expect(getByText('All indices')).toBeInTheDocument();
    });
    test('should display selected detector state and index options', () => {
      const updatedProps = {
        ...defaultProps,
        selectedDetectorStates: [DETECTOR_STATE.DISABLED],
        selectedIndices: ['test_index'],
      };
      const { getByText } = render(<ListControls {...updatedProps} />);
      expect(getByText(DETECTOR_STATE.DISABLED)).toBeInTheDocument();
      expect(getByText('test_index')).toBeInTheDocument();
    });
    test('should call onIndexSearchChange when searching in index filter', () => {
      const { getAllByTestId } = render(<ListControls {...defaultProps} />);
      userEvent.type(getAllByTestId('comboBoxSearchInput')[1], 'Testing');
      expect(defaultProps.onSearchIndexChange).toHaveBeenCalledTimes(7);
    });
    test('should display multiple selected detector state and index options', () => {
      const updatedProps = {
        ...defaultProps,
        selectedDetectorStates: [DETECTOR_STATE.DISABLED, DETECTOR_STATE.INIT],
        selectedIndices: ['test_index_1', 'test_index_2'],
      };
      const { getByText } = render(<ListControls {...updatedProps} />);
      expect(getByText(DETECTOR_STATE.DISABLED)).toBeInTheDocument();
      expect(getByText(DETECTOR_STATE.INIT)).toBeInTheDocument();
      expect(getByText('test_index_1')).toBeInTheDocument();
      expect(getByText('test_index_2')).toBeInTheDocument();
    });
  });
});
