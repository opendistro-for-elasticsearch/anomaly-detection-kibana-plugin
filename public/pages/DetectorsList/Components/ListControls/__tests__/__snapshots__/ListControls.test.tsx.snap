// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListControls /> spec Empty results renders component with empty message 1`] = `
<div
  class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiFlexGroup--responsive"
>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
    style="width: 40%;"
  >
    <div
      class="euiFormControlLayout euiFormControlLayout--fullWidth"
    >
      <div
        class="euiFormControlLayout__childrenWrapper"
      >
        <input
          class="euiFieldSearch euiFieldSearch--fullWidth"
          data-test-subj="detectorListSearch"
          placeholder="Search"
          type="search"
          value=""
        />
        <div
          class="euiFormControlLayoutIcons"
        >
          <span
            class="euiFormControlLayoutCustomIcon"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
              focusable="false"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      class="euiComboBox euiComboBox--fullWidth"
      role="combobox"
    >
      <div
        class="euiFormControlLayout euiFormControlLayout--fullWidth"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <div
            class="euiComboBox__inputWrap euiComboBox__inputWrap--fullWidth euiComboBox__inputWrap-isClearable"
            data-test-subj="comboBoxInput"
            tabindex="-1"
          >
            <p
              class="euiComboBoxPlaceholder"
            >
              All detector states
            </p>
            <div
              class="euiComboBox__input"
              style="font-size: 14px; display: inline-block;"
            >
              <input
                data-test-subj="comboBoxSearchInput"
                id="selectedDetectorStates"
                role="textbox"
                style="box-sizing: content-box; width: 2px;"
                value=""
              />
              <div
                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
              />
            </div>
          </div>
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <button
              aria-label="Open list of options"
              class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
              data-test-subj="comboBoxToggleListButton"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                focusable="false"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      class="euiComboBox euiComboBox--fullWidth"
      role="combobox"
    >
      <div
        class="euiFormControlLayout euiFormControlLayout--fullWidth"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <div
            class="euiComboBox__inputWrap euiComboBox__inputWrap--fullWidth euiComboBox__inputWrap-isClearable"
            data-test-subj="comboBoxInput"
            tabindex="-1"
          >
            <p
              class="euiComboBoxPlaceholder"
            >
              All indices
            </p>
            <div
              class="euiComboBox__input"
              style="font-size: 14px; display: inline-block;"
            >
              <input
                data-test-subj="comboBoxSearchInput"
                id="selectedIndices"
                role="textbox"
                style="box-sizing: content-box; width: 2px;"
                value=""
              />
              <div
                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
              />
            </div>
          </div>
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <button
              aria-label="Open list of options"
              class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
              data-test-subj="comboBoxToggleListButton"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                focusable="false"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <div
      class="euiPagination"
      data-test-subj="detectorPageControls"
      role="group"
    >
      <button
        aria-label="Previous page"
        class="euiButtonIcon euiButtonIcon--text"
        data-test-subj="pagination-button-previous"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
          focusable="false"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        />
      </button>
      <button
        aria-label="Page 1 of 5"
        class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiPaginationButton euiPaginationButton--hideOnMobile"
        data-test-subj="pagination-button-0"
        type="button"
      >
        <span
          class="euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            1
          </span>
        </span>
      </button>
      <button
        aria-label="Page 2 of 5"
        class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiPaginationButton euiPaginationButton-isActive euiPaginationButton--hideOnMobile"
        data-test-subj="pagination-button-1"
        type="button"
      >
        <span
          class="euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            2
          </span>
        </span>
      </button>
      <button
        aria-label="Page 3 of 5"
        class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiPaginationButton euiPaginationButton--hideOnMobile"
        data-test-subj="pagination-button-2"
        type="button"
      >
        <span
          class="euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            3
          </span>
        </span>
      </button>
      <button
        aria-label="Page 4 of 5"
        class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiPaginationButton euiPaginationButton--hideOnMobile"
        data-test-subj="pagination-button-3"
        type="button"
      >
        <span
          class="euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            4
          </span>
        </span>
      </button>
      <button
        aria-label="Page 5 of 5"
        class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiPaginationButton euiPaginationButton--hideOnMobile"
        data-test-subj="pagination-button-4"
        type="button"
      >
        <span
          class="euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            5
          </span>
        </span>
      </button>
      <button
        aria-hidden="true"
        class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiPaginationButton euiPaginationButton-isPlaceholder euiPaginationButton--hideOnMobile"
        disabled=""
        type="button"
      >
        <span
          class="euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            <span>
              â€¦
            </span>
          </span>
        </span>
      </button>
      <button
        aria-label="Jump to the last page, number 10"
        class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiPaginationButton euiPaginationButton--hideOnMobile"
        type="button"
      >
        <span
          class="euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            10
          </span>
        </span>
      </button>
      <button
        aria-label="Next page"
        class="euiButtonIcon euiButtonIcon--text"
        data-test-subj="pagination-button-next"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
          focusable="false"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        />
      </button>
    </div>
  </div>
</div>
`;
